File: time_synchronization/python/ptp_sync.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
#
# Copyright (c) 2025 SICK AG, Waldkirch
#
# SPDX-License-Identifier: Unlicense

import argparse
from python_base.Control import Control
from shared.python.devices_config import get_device_config
from datetime import datetime, timezone
import time

def run_PTP_demo(ip_address: str):
    # tag::open_control_channel[]
    device_control = Control(ip_address, cola_protocol, control_port)
    device_control.open()
    # end::open_control_channel[]

    # tag::timestamp_before[]
    device_timestamp_ms = device_control.getDeviceTime()
    device_time = datetime.fromtimestamp(device_timestamp_ms / 1000, tz=timezone.utc).strftime('%Y-%m-%d %H:%M:%S')
    print(f"Device time before time synchronization: {device_time}")
    # end::timestamp_before[]

    print(f"PTP State: {device_control.getPtpState()}")

    # tag::wait_for_userinput[]
    # Ask user about PTP Master
    print("Running PTP Master is connected to the network? [Y/N]")
    user_input = input().strip().upper()
    # end::wait_for_userinput[]

    if user_input == 'Y':
        # tag::login[]
        device_control.login(Control.USERLEVEL_SERVICE, 'CUST_SERV')
        # end::login[]
        
        # tag::ptp_setting[]
        device_control.setPtpMode("SLAVE")
        print("Set PTP Mode: SLAVE")
        # end::ptp_setting[]

        # tag::enable_ptp[]
        device_control.setTimeSyncMode('PTP')
        print("Enabled PTP time synchronization")
        # end::enable_ptp[]
        device_control.logout()

        # tag::sleep[]
        print("Waiting for 5 seconds to be on the safe side.")
        time.sleep(5)
        # end::sleep[]
        print(f"PTP State: {device_control.getPtpState()}")

        # tag::timestamp_after[]
        device_timestamp_ms = device_control.getDeviceTime()
        device_time = datetime.fromtimestamp(device_timestamp_ms/1000, tz=timezone.utc).strftime('%Y-%m-%d %H:%M:%S')
        print(f"Device time after time synchronization: {device_time}")
        # end::timestamp_after[]
        
    # tag::disable_ptp[]
    device_control.login(Control.USERLEVEL_SERVICE, 'CUST_SERV')
    device_control.setTimeSyncMode('NONE')
    print("Disabled PTP (timeSyncMode)")
    # end::disable_ptp[]
    
    # tag::logout_and_close[]
    device_control.logout()
    device_control.close()
    # end::logout_and_close[]
    print("Logged out. Closed connection.")

if __name__ == "__main__":
    parser = argparse.ArgumentParser(
        description="This script demonstrates how to synchronize the device time with PTP.")
    parser.add_argument('-i', '--ip_address', required=False, type=str,
                        default="192.168.1.10", help="The ip address of the device.")
    args = parser.parse_args()

    cola_protocol, control_port, _ = get_device_config("Visionary-T Mini")

    run_PTP_demo(args.ip_address)

File: time_synchronization/cpp/ptp_sync.cpp
//
// Copyright (c) 2024 SICK AG, Waldkirch
//
// SPDX-License-Identifier: Unlicense

#include <algorithm>
#include <cinttypes>
#include <cstdint>
#include <cstdio>
#include <string>

#include <iostream>
#include <sstream>
#include <thread>

#include <sick_visionary_cpp_base/CoLaParameterWriter.h>
#include <sick_visionary_cpp_base/VisionaryControl.h>
#include <sick_visionary_cpp_base/VisionaryType.h>

#include "PtpHelpers.h"
#include "TimeStampHelper.h"
#include "exitcodes.h"

static ExitCode runPtpDemo(visionary::VisionaryType visionaryType, const std::string& ipAddress)
{
  using namespace visionary;

  VisionaryControl visionaryControl(visionaryType);

  // Connect to devices control channel of the visionary device
  if (!visionaryControl.open(ipAddress))
  {
    std::fprintf(stderr, "Open control connection failed.\n");
    return ExitCode::eCommunicationError;
  }
  std::printf("Opened connection to Visionary device: %s\n", ipAddress.c_str());

  // Get device time before synchronization
  std::printf("Device time before time synchronization: ");
  printDeviceTime(visionaryControl);

  // Read PTP state before enabling PTP
  std::printf("PTP State: %s\n", getPtpState(visionaryControl).c_str());

  // Ask user about PTP Master
  std::string userInput;
  std::printf("Working PTP Master is connected to the network? [Y/N]: ");
  std::getline(std::cin, userInput);

  // Convert to uppercase for consistency
  std::transform(userInput.begin(), userInput.end(), userInput.begin(), ::toupper);

  if (userInput == "Y" || userInput == "y")
  {
    // Login
    if (!visionaryControl.login(IAuthentication::UserLevel::AUTHORIZED_CLIENT, "CLIENT"))
    {
      std::fprintf(stderr, "Login failed.\n");
      return ExitCode::eAuthenticationError;
    }
    
    // PTP settings
    // Set device to PTP Slave
    CoLaCommand setPtpModeCommmand = CoLaParameterWriter(CoLaCommandType::WRITE_VARIABLE, "ptpMode")
                                       .parameterUSInt(2)
                                       .build(); // 0 = AUTO, 1 = MASTER, 2 = SLAVE
    CoLaCommand setPtpModeCommmandResponse = visionaryControl.sendCommand(setPtpModeCommmand);
    if (setPtpModeCommmandResponse.getError() != CoLaError::OK)
    {
      std::fprintf(stderr, "Set PTP Mode failed.\n");
      return ExitCode::eParamError;
    }
    else
    {
      std::printf("Set PTP Mode: SLAVE\n");
    }

    /// Enable PTP
    CoLaCommand enablePtpCommand =
      CoLaParameterWriter(CoLaCommandType::WRITE_VARIABLE, "timeSyncMode").parameterUSInt(2).build(); // 0 = NONE, 1 = Ptp, 2 = PTP 
    CoLaCommand enablePtpCommandResponse = visionaryControl.sendCommand(enablePtpCommand);
    if (enablePtpCommandResponse.getError() != CoLaError::OK)
    {
      std::fprintf(stderr, "Enable PTP time sync failed.\n");
      return ExitCode::eParamError;
    }
    else
    {
      std::printf("Enabled PTP (timeSyncMode)\n");
    }
    visionaryControl.logout();

    std::this_thread::sleep_for(std::chrono::milliseconds(1000));
    std::printf("PTP State: %s\n", getPtpState(visionaryControl).c_str());

    std::printf("Waiting 5 seconds for the PTP state update.\n");
    std::this_thread::sleep_for(std::chrono::milliseconds(5000));
    
    // Get device time after synchronization
    std::printf("Device time after time synchronization\n");
    printDeviceTime(visionaryControl);
    }

    // Disable PTP
    if (!visionaryControl.login(IAuthentication::UserLevel::AUTHORIZED_CLIENT, "CLIENT"))
    {
      std::fprintf(stderr, "Login failed.\n");
      return ExitCode::eAuthenticationError;
    }
    CoLaCommand disablePtpCommand =
        CoLaParameterWriter(CoLaCommandType::WRITE_VARIABLE, "timeSyncMode").parameterUSInt(0).build();
    CoLaCommand disablePtpCommandResponse = visionaryControl.sendCommand(disablePtpCommand);
    if (disablePtpCommandResponse.getError() != CoLaError::OK)
    {
      std::fprintf(stderr, "Disable PTP time sync failed\n");
      return ExitCode::eParamError;
    }
    else
    {
      std::printf("Disabled PTP (timeSyncMode)\n");
    }

    visionaryControl.logout();
    visionaryControl.close();
    std::printf("Logout and closed connection.\n");
    return ExitCode::eOk;
}

int main(int argc, char* argv[])
{
  std::string              deviceIpAddr("192.168.1.10");
  visionary::VisionaryType visionaryType(visionary::VisionaryType::eVisionaryTMini);

  bool showHelpAndExit = false;

  ExitCode exitCode = ExitCode::eOk;

  for (int i = 1; i < argc; ++i)
  {
    std::istringstream argstream(argv[i]);

    if (argstream.get() != '-')
    {
      showHelpAndExit = true;
      exitCode        = ExitCode::eParamError;
      break;
    }
    switch (argstream.get())
    {
      case 'h':
        showHelpAndExit = true;
        break;
      case 'i':
        argstream >> deviceIpAddr;
        break;
      default:
        showHelpAndExit = true;
        break;
    }
  }

  if (showHelpAndExit)
  {
    std::cout << argv[0] << " [option]*"
              << "\n";
    std::cout << "where option is one of"
              << "\n";
    std::cout << "-h          show this help and exit"
              << "\n";
    std::cout << "-i<IP>      connect to the device with IP address <IP>; default is " << deviceIpAddr << "\n";
    return static_cast<int>(exitCode);
  }

  exitCode = runPtpDemo(visionaryType, deviceIpAddr);

  std::cout << "exit code " << static_cast<int>(exitCode) << "\n";

  return static_cast<int>(exitCode);
}
